Code/Framework/AzCore/AzCore/JSON/rapidjson.h

```diff
@@ -22,38 +19,48 @@
 #define RAPIDJSON_HAS_CXX11_TYPETRAITS 1
 #define RAPIDJSON_HAS_CXX11_RVALUE_REFS 1
 
-namespace rapidjson_ly_internal
+// note for future maintainers:
+// RapidJSON allows you to customize the default allocator type for documents by overriding with RAPIDJSON_DEFAULT_ALLOCATOR.
+// However, customizing the default type of rapidjson::Document and rapidjson::Value in that way will cause incompatibilities
+// with external software libraries that use unmodified RapidJSON.   It is not thus recommended
+// to actually the value of RAPIDJSON_DEFAULT_ALLOCATOR or redefine what a rapidjson::Document itself is.
+// The other way to get a 'custom' allocator installed in RapidJSON is to make your own Document/Value/Pointer type, since its templated
+// on allocator type - for example, implement your own allocator and do
+// using MyDocument = rapidjson::GenericDocument<rapidjson::UTF8<char>, MyAllocator, MyAllocator> and then use MyDocument in your code
+// instead of RapidJSON::Document.  Note GenericDocument<> customized must use customized GenericValue<> and GenericPointer<>
+// or else you haven't acutally plugged into all the places where memory is allocated and will lose some.
+// 
+// Unfortunately, rapidjson::Pointer can't deal with these customized values or documents due to a bug in the code where functions
+// like Set/Get/Create use the default template parameter instead of deducing it:
+// ie, they are set up like somefunction<T>(GenericValue<char> val) which will only accept default GenericValue<char, CrtAllocator>.
+// correct form: somefunction<T>(GenericValue<char, T::AllocatorType> val, which would allow MyValue<char, MyAllocator> to be passed in.
+// thus, it is not recommended to use customized documents, values, or pointers either, (until that bug is fixed).
+
+// this leaves just one avenue for customization of memory of Rapidjson - overriding the macros RAPIDJSON_xxxx (new,delete, malloc, realloc..)
```

Comment:
Should we also add in ifdef checks to make sure the macros aren't already defined before defining them?

So uses can technically avoid defined there own allocators through compiler definitions instead?

Reply:
I can do that, yes.  Its a good idea, since that's what rapidjson already does also.

Reply:
Should we use our own macros or adopt rapidjsons ones?  Hmm.

Reply:
I vote for using the rapidjson macros.

The benefits are that there is only one set of macros that users need to have knowledge about and they could lookup that information in the rapidjson documentation.
Furthermore it helps avoid any discovery issues with determining that we supply macros for overriding the allocator.